combustion-android-ble/src/main/java/inc/combustion/framework/ble/BleDevice.kt

```diff
@@ -0,0 +1,253 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package inc.combustion.framework.ble
+
+import android.bluetooth.BluetoothAdapter
+import android.os.SystemClock
+import android.util.Log
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import com.juul.kable.*
+import inc.combustion.framework.LOG_TAG
+import inc.combustion.framework.service.DeviceConnectionState
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.onCompletion
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Base class for Combustion Devices.
+ *
+ * @property mac Bluetooth MAC address of device.
+ * @property owner Owner of the instance's coroutine scope.
+ * @constructor Creates a new Device with the specified MAC.
+ *
+ * @param adapter Android BluetoothAdapter.
+ */
+internal open class BleDevice (
+    private val mac: String,
+    protected val owner: LifecycleOwner,
+    var advertisingData: LegacyProbeAdvertisingData,
+    adapter: BluetoothAdapter
+) {
+    companion object {
+        private const val PROBE_REMOTE_RSSI_POLL_RATE_MS = 2000L
+        private const val PROBE_IDLE_TIMEOUT_POLL_RATE_MS = 1000L
+        private const val PROBE_IDLE_TIMEOUT_MS = 15000L
+        const val DISCONNECT_TIMEOUT_MS = 500L
+
+        private const val DEV_INFO_SERVICE_UUID_STRING = "0000180A-0000-1000-8000-00805F9B34FB"
+        private val SERIAL_NUMBER_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A25-0000-1000-8000-00805F9B34FB"
+        )
+        private val FW_VERSION_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A26-0000-1000-8000-00805F9B34FB"
+        )
+        private val HW_REVISION_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A27-0000-1000-8000-00805F9B34FB"
+        )
+    }
+
+    val jobManager = JobManager()
+    protected var peripheral: Peripheral =
+        owner.lifecycleScope.peripheral(adapter.getRemoteDevice(mac)) {
+            logging {
+                /* The following enables logging in Kable
+                // engine = SystemLogEngine
+                // level = Logging.Level.Events
+                // format = Logging.Format.Multiline
+                // data = Hex
+                 */
+            }
+        }
+
+    protected val connectionMonitor = IdleMonitor()
+    val remoteRssi = AtomicInteger(0)
```

Comment:
Spoke with Matt about this--please put a `TODO` in here to make private when advertising data injection's addressed.

Reply:
I hope this captures the design intent for when scanning is integrated.

```
    // TODO: This variable should ideally be scoped as private.  
    //  This should be possible when this class is able handle and process advertising
    //  packets directly from the device scanner.
```